### ISSUE TITLE
vertical-horizontal wrapper not working properly

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#414]
In GitLab by @ editomas on Jun 11, 2019, 14:29

@mcastril @dbeltrankyl 

there are cases in which the horizional-vertical wrapper starts to behave strangely (not clear yet to me if this happens after a job falure or systematically). 

here are three issues (a, b,c) when I was running 2 starting dates (with the crossing-date strategy): from 1 Jan and 1 Jul for 1 month each in exp a1yw.

- first problematic wrapper:

it has created this wrapper job with only Jul SIMs and DA (not Jan)

 ```
ls -l /gpfs/scratch/pr1ehu00/pr1ehu15/a1yw/LOG_a1yw/a1yw_ASThread_15600045392833_15552_195.cmd
-rwxrwxr-x 1 pr1ehu15 pr1ehu00 11049 jun  8 16:45 /gpfs/scratch/pr1ehu00/pr1ehu15/a1yw/LOG_a1yw/a1yw_ASThread_15600045392833_15552_195.cmd
```

or

`/esarchive/autosubmit/a1yw/tmp> vi a1yw_ASThread_15600045392833_15552_195.cmd`

I think the log is this one:

`/esarchive/autosubmit/a1yw/tmp> vi 20190606_175701_run.log`

**issue a) why has not included Jan?**
 17 Jan was free to be run since I can see the all the SIMs before were completed:

```
pr1ehu15@login1:/gpfs/scratch/pr1ehu00/pr1ehu15/a1yw/LOG_a1yw> ls -lrt *201201*_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_007_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_001_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_011_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_009_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_006_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_004_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_010_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_002_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_005_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_003_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 13:56 a1yw_20120101_008_17_SIM_COMPLETED
-rw-r--r-- 1 pr1ehu15 pr1ehu00 0 jun  8 14:26 a1yw_20120101_000_17_SIM_COMPLETED
```

**issue b) this is wasting a lot of resources** each time the Da is run: the first chunk is made of 12 SIMs the second chunk is made only of a DA job (which needs only 2500 cores instead of the 15552 reserved by the wrapper)

hence it finishes to run the full July and then prepares a wrapper for the rest of Jan dates not completed yet (i.e. from the 17 Jan) and here it comes the problem n.2

-second problematic wrapper

it prepares a wrapper asking for 2500 cores:

`/gpfs/scratch/pr1ehu00/pr1ehu15/a1yw/LOG_a1yw> vi a1yw_ASThread_15600668827226_2500_183.cmd`
or
`/esarchive/autosubmit/a1yw/tmp> vi a1yw_ASThread_15600668827226_2500_183.cmd`

I think the run.log is this one:

`/esarchive/autosubmit/a1yw/tmp> vi 20190609_095333_run.log`

the first chunk is a DA job that needs 2500 cores, the second chunk is 12 SIMs, where each SIM needs 1296 cores!

**issue c) so this wrapper will never work.**

It was one of those wrappers that got stuck because of the recent module issue #410, but I've run the wrapper manually in HPC just to prove that it would run  `a1yw_20120701_17_DA.cmd` and then fail.)

To summarize:

a) the main problem is that if it looses the ability to cross dates. not sure if it is as soon as there is a problem (I could not even find out if there was a failure) or systematically.

b) it creates wrappers that waste a lot of resources (if the SIMs go in the first chuck)

c) it creates wrappers that will never work (if DA goes in the first chunk)

is there a fix that could be applied to force in the first place the crossing of dates?

for info, @orioljorba @f-macchia @gmontanepinto

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 11, 2019, 15:59

a plot confirming the above:
`/usr/bin/eog /esarchive/autosubmit/a1yw/plot/a1yw_20190611_1546.svg`

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 11, 2019, 18:07

Hi @ editomas . It's a big workflow and not easy to see. But from what I saw I think the main issue is that is not crossing the dates right? It is doing January (SIM+DA) in one wrapper and July (SIM+DA) in the other.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 11, 2019, 18:33

@mcastril yes one issue is that it does not cross date (issue (a) above), but not from the beginning, initially it does well. The first issue met is actually that it starts to produce wrappers of the situation (c: not enough cores for all chunks) and (a: not crossing dates). 

Looking at all the AST.com files here it is what I see:

it starts well with crossing the dates, for both months in one unique wrapper of 18052 cores (number correct for 12 SIMs+DA) :

`/esarchive/autosubmit/a1yw/tmp/a1yw_ASThread_15598367078156_18052_794.cmd`

then something goes wrong, maybe a failure (?), and produces this wrapper of 1296 cores:

`/esarchive/autosubmit/a1yw/tmp/a1yw_ASThread_15599954661753_1296_184.cmd`

which, not only does not cross the dates, but clearly can do only the first chunk (1 SIM), it will already not have enough cores for the second chunk (1 DA) or the third chuck (12 SIM):

`['a1yw_20120101_000_17_SIM.cmd'], ['a1yw_20120101_17_DA.cmd'], ['a1yw_20120101_010_18_SIM.cmd', 'a1yw_20120101_008_18_SIM.cmd', 'a1yw_20120101_007_18_SIM.cmd', 'a1yw_20120101_006_18_SIM.cmd', 'a1yw_20120101_005_18_SIM.cmd', 'a1yw_20120101_004_18_SIM.cmd', 'a1yw_20120101_009_18_SIM.cmd', 'a1yw_20120101_002_18_SIM.cmd', 'a1yw_20120101_001_18_SIM.cmd', 'a1yw_20120101_000_18_SIM.cmd', 'a1yw_20120101_003_18_SIM.cmd', 'a1yw_20120101_011_18_SIM.cmd']`

(please correct me if I'm using wrongly the word "chunk" in my description.) Somehow it takes into account only the cores of the first chunk that has to do, while the rest of the chunks might instead need a bigger or smaller number of cores. 

then it produces the two other problematic wrapper mentioned at the beginning of the issue.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 11, 2019, 19:10

@mcastril additional info: I've discovered in `/esarchive/autosubmit/a1yw/tmp/20190606_175701_run.log` what has caused the first (and only good) wrapper to fail:

```
2019-06-08 14:03:48,493 Command find /gpfs/scratch/pr1ehu00/pr1ehu15/a1yw/LOG_a1yw  -name \*a1yw_20120101_000_17_SIM_COMPLETED in mn4_prace_cp successful with out message:
[ERROR] 2019-06-08 14:03:48,493 Job a1yw_20120101_000_17_SIM inside wrapper a1yw_ASThread_15598367078156_18052_794 is running for longer than its wallclock! Setting to FAILED...
```

it must have been machine underperormance. 

the problem is that after the failure of 1 SIM is not able anymore to build good wrappers. After the failure of a1yw_20120101_000_17_SIM.cmd, as I said above, it builds a 1296 cores wrapper with the following chunks that won't enough all enough cores and all from January:

`scripts = [['a1yw_20120101_000_17_SIM.cmd'], ['a1yw_20120101_17_DA.cmd'], ['a1yw_20120101_010_18_SIM.cmd', 'a1yw_20120101_008_18_SIM.cmd', 'a1yw_20120101_007_18_SIM.cmd', 'a1yw_20120101_006_18_SIM.cmd', 'a1yw_20120101_005_18_SIM.cmd', 'a1yw_20120101_004_18_SIM.cmd', 'a1yw_20120101_009_18_SIM.cmd', 'a1yw_20120101_002_18_SIM.cmd', 'a1yw_20120101_001_18_SIM.cmd', 'a1yw_20120101_000_18_SIM.cmd', 'a1yw_20120101_003_18_SIM.cmd', 'a1yw_20120101_011_18_SIM.cmd'], ['a1yw_20120101_18_DA.cmd'] etc...`

I know it is a bit complicated, but ideally it should run first only the failed SIM (or whatever jobs have failed) and then it could go back to its crossing-date strategy without chances that to mess up with number of cores and not crossing.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 12, 2019, 13:19

Hi @ editomas . So, to recap all the information that you provided, and our discussion now, it seems the vertical-horizontal wrapper has two bugs:

*  The first bug is that when one SIM fails (or the DA) it starts a new wrapper putting only the failed SIM (or the DA) in the first top, instead of running the isolate job outside the wrapper and starting a new wrapper with 12 SIM + 1 DA.

*  The second bug is that it calculates the total number of processes needed taking into account only the first row. For instance, if only one SIM is run at the beginning, it will only allocate enough cores for one SIM, and when the second steps starts it will fail because it is actually running 1DA or 12 SIM...

We should implement a mechanism that makes Autosubmit not to wrap rows of different size in the same job. SIM and DA that cannot be grouped with other SIM and DA should be run separately, and the wrapper should contain the same workload in all the rows. This would fix both bugs at the same time.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 2, 2019, 17:22

Hi there @ editomas @mcastril @f-macchia ,

I did a somewhat big commit for 3.12.1b(that will be online tomorrow morning, since I leave a large test today and will check it tomorrow)  which should solve this issue alongside some other issues that I discovered one related is -cw which now should work always and is useful to check how the wrapper will look like with the actual job status.

I'll notify you tomorrow on this issue when it is online.

(3.12.0b won't receive this commit)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 2, 2019, 17:23

mentioned in commit 91fde47ec0a3e404db7b15a6bdebc426048feba2

**Comment by kinow** [comment]
In GitLab by @f-macchia on Jul 3, 2019, 12:04

@dbeltrankyl I'm sorry but no new experiments are planned for me, so I cannot test the new changes.

@ editomas please let me know if plans change.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 3, 2019, 15:13

Hi, 
Thanks @f-macchia for the info, in any case, it is updated already(after a few more changes) however I think none of auto-monarch is using this 3.12.1b version then so, @ editomas in case  you want to test it keep in mind that it is less reliable than 3.12.0b

**Comment by kinow** [comment]
In GitLab by @f-macchia on Jul 16, 2019, 16:10

Happy to do the testing, but I need experiments to run.

In the last DustClim meeting we discussed about one month of simulation to test the model before running 14 year. Can I run this experiment to test these developments?

@ editomas @orioljorba @ sbasart what do you think?

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 16, 2019, 20:36

@f-macchia yes we have a simulation to be run. we said today to finish 2012 saving only a basic set of variables. 

For this I will continue the exp a20u from where it is now, but I'm also missing the pair Jan/Jul (the exp is using two starting dates).

If you want to make a test with autosubmit, you could run this pair yourself. 

I will have to tell you the exact starting dates (looking at the available history files), number of chunks (so that they overlap for some days with current available results to have continuity), the csv for the reduce.   

if you want to do it, let me know and I write these details. thanks

**Comment by kinow** [comment]
In GitLab by @f-macchia on Jul 17, 2019, 09:52

Yes, sure! Testing with a real case is the best option!

Let me know the details, I'll run this test with autosubmit 3.12.1b.

We should also take into account the available hours on prace, we have not so much hour for "testing" and we need to proceed with the final run. But we can discuss that in the next meeting.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jul 19, 2019, 18:20

@f-macchia here details for your test.

  for the exp.conf:

```
DATELIST = 2011123100  20120629
CHUNKINI=1
NUMCHUNKS = 38
```

  for the proj.conf:

`DA_FOLDER_HPC    = /gpfs/scratch/bsc32/bsc32137/PREOBS_2019/obs_deepblue_coarse_dod_balance_05_flatter/`

`SPINUP_FOLDER = /gpfs/archive/bsc32/esarchive/exp/monarch/a1zw/original_files/`

please see the REDUCE_CSV_FILE here  /shared/earth/AtmosphericComposition/WG_DATA_ASSIMILATION/dustclim/
you could use:

(a) `REDUCE_CSV_FILE = monarch_reduce_DUST_DA_odco_odtot_pm10_solar.csv`  (this is what I'm using for the rest of the months)

or

(b) `REDUCE_CSV_FILE =monarch_reduce_DUST_DA_dustclim.csv`  (full set of variables to be finally stored)

or

(c) `REDUCE_CSV_FILE =monarch_reduce_DUST_DA_dustclim_min.csv` (minimum set of variables to be finally stored, if we go for this scenario)

  for the above 3 versions the corresponding `proj/auto-monarch/sources/nmmb-monarch/MODEL/NAMELISTS/solver_state.DUST_DA_NPS.txt` has to be equal (hence manually copied, cannot be set at configuration level) to:

(a) `solver_state.DUST_DA_NPS_odco_odtot_pm10_solar.txt`

(b) `solver_state.DUST_DA_NPS_dustclim.txt`

(c) `solver_state.DUST_DA_NPS_dustclim_min.txt`

the above three files are also in `/shared/earth/AtmosphericComposition/WG_DATA_ASSIMILATION/dustclim/`

I hope I did not forget anything. Before running it we can compare to the configuration of exp a20u which is similar.

**Comment by kinow** [comment]
In GitLab by @f-macchia on Jul 30, 2019, 12:40

Testing autosubmit/3.12.1b-foss-2015a-Python-2.7.9 with horizontal-vertical wrapper (expid a24p).

I receive this error:
```
16619 of 16644 jobs remaining (12:36)
Jobs ready for marenostrum4: 25
[WARNING] Command /gpfs/scratch/pr1ehu00/pr1ehu06/a24p/LOG_a24p/submit_marenostrum4.sh in mn4 warning: sbatch: error: Invalid numeric value "0" for number of tasks.

[ERROR] marenostrum4 submission failed
[CRITICAL] Unhandled exception on Autosubmit: list index out of range
Traceback (most recent call last):
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/autosubmit.py", line 418, in parse_args
    return Autosubmit.run_experiment(args.expid, args.notransitive)
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/autosubmit.py", line 1076, in run_experiment
    if Autosubmit.submit_ready_jobs(as_conf, job_list, platforms_to_test, packages_persistence):
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/autosubmit.py", line 1166, in submit_ready_jobs
    job.id = str(jobs_id[i])
IndexError: list index out of range
```

Log (with DEBUG option) in `/esarchive/autosubmit/a24p/tmp/ASLOGS/20190730_122926_run.log`.

I think it's because the total number of required cores is 0. 

@dbeltrankyl please could you check if I'm doing something wrong in the configuration of the wrapper?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 30, 2019, 13:21

hi @f-macchia , 

I'll debug autosubmit with your current configuration, for now, the only thing that could I think of failure is dependencies = True, it is necessary for your experiment? 

I'll work asap when I finish another issue.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 30, 2019, 13:23

Also, now you're including SIM&DA&REDUCE , in the previous workflow what I saw was only 
```
JOBS_IN_WRAPPER = SIM&DA
JOBS_IN_WRAPPER = SIM&DA REDUCE
```
So I'll double check that SIM&DA&REDUCE is working properly

**Comment by kinow** [comment]
In GitLab by @f-macchia on Jul 30, 2019, 15:02

I commented "DEPENDENCIES = True", but I have the same error.

**Comment by kinow** [comment]
In GitLab by @f-macchia on Jul 30, 2019, 15:25

Same error also using `JOBS_IN_WRAPPER = SIM&DA REDUCE`.

However, keep in mind that the final configuration will be `JOBS_IN_WRAPPER = SIM&DA&REDUCE`.

I leave a24p unchanged with the final configuration that we need to run.

Thanks!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 30, 2019, 15:27

Sorry for the issue and thanks for the update, I'll notify you when I find the bug

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 1, 2019, 13:11

hi @f-macchia , 

Apologizes for the delay sounds like an excuse but yesterday in WFH my pc crashed  had to reformat it and then It wasn't detecting the VPN while feeling myself ill ( great day ) 

It appears to be working properly now in the dev version, I merge this in 3.12.1b together with a few merge requests from Wilmer( setstatus mainly) I test altogether and update the module.

I guess it will be available at 15:00pm ~ but will notify you.

**Comment by kinow** [comment]
In GitLab by @f-macchia on Aug 1, 2019, 14:53

Don't worry, it can happen.. :D

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 1, 2019, 15:14

hi @f-macchia ,

I did some test with lesser cores and using inspect and run commands, 

It should works now in 3.12.1b( version updated already), however dependencies = True should be debugged in a real experiment since it still is untested more than in dummies. 

Fixes to total procs and also now supports +2jobs with & (it behavior was wrong )

This upload includes #374,#372,#386,

Worth to mention for your particular case that if you want to use for example 

`setstatus -t COMPLETED  -fc "[ 20150527 [ 000 [1] 001 [1] 002 [1] 003 [1] 004 [1 ] 005 [1] 006 [1] 007 [1] 008 [1 ] 009 [1] 010 [1] 011 [1]  ] ],SIM" a257 -s`

(sets to complete the first sims to enable double-date, )

it was swaped to `-fct ` flag

**Comment by kinow** [comment]
In GitLab by @f-macchia on Aug 1, 2019, 17:49

The wrapper is correctly created and submitted! Thanks!!!!

I'm submitting jobs without using dependencies. I'll try dependencies in an other test. I'll let you know..

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 5, 2019, 11:26

I'm glad that it worked, closing the issue if there are any problem feel free to reopen it

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 5, 2019, 11:26

closed
